### ISSUE TITLE
IndexedRelations

### ISSUE THREAD

**Issue by LTLA** [Bioconductor/Contributions#1114]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/LTLA/IndexedRelations

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/

**Comment by LTLA** [comment]
To put this into context for whoever reviews this: I should mention that this particular package is Act 1 of a 3-part epic to refactor the Bioconductor infrastructure for genomic interactions.

**Comment by lshep** [comment]
We are working out some kinks with teh SPB relating to the release. Should be resolved soon ...

**Comment by hpages** [comment]
Hi Aaron,

Thanks for this new submission. Congrats for taking advantage of things like `parallelSlotNames`, `bindROWS`, and `pcompare` to get many core operations work out-of-the-box on your Vector derivatives! Before I proceed into a detailed review, can you please look into the following issues?

```
rel1 <- IndexedRelations(list(l=2), list(letters))
rel2 <- IndexedRelations(list(l=1), list(letters[-1]))
pcompare(rel1, rel2)
# Error in (function (classes, fdef, mtable)  : 
#   unable to find an inherited method for function ‘mcols’ for signature ‘"character"’
```

```
rel3 <- IndexedRelations(list(a=11:13, A=1:3), list(letters, LETTERS[1:5]))
rel4 <- IndexedRelations(list(a=7:8, A=4:5), list(letters, LETTERS[3:7]))
c(rel3, rel4)
# Error in (function (classes, fdef, mtable)  : 
#   unable to find an inherited method for function ‘mcols’ for signature ‘"character"’
```

```
rel5 <- IndexedRelations(list(A=7:8, a=5:4), list(letters, LETTERS[1:5]))
c(rel3, rel5)
# Error in validObject(ans) : invalid class “IndexedRelations” object: 
#     column 2 of 'partners' contains out-of-bounds indices
```

Probably not intended to work that way:
```
rel6 <- IndexedRelations(list(A=5:4, a=1:2), list(letters, LETTERS[1:5]))
c(rel3, rel6)
# IndexedRelations object with 5 relations and 0 metadata columns:
#                 a           A
#       <character> <character>
#   [1]           k           A
#   [2]           l           B
#   [3]           m           C
#   [4]           a           E
#   [5]           b           D
```

FWIW I see your IndexedRelations as a multi-dimensional generalization of factors with a set of levels along each dimension. Unlike with factors, the levels here can be any vector-like object and can contain duplicates. Personally I see the fact that they can contain duplicates as a complication that is maybe unnecessary. Do you really need that flexibility? Otherwise you could just forbid them (the constructor could either remove them and adjust the indexes, or simply reject them). I think this would allow a more canonical way to combine levels in the context of a binary operation. For example, when doing things like `c(rel1, rel2)` or `rel1[i] <- rel2`, the levels of the result would be obtained by appending to the levels in `rel1` the levels in `rel2` that are not in `rel1`. Note that this is technically a union but unfortunately you wouldn't be able to use `union()` because it deviates from the standard meaning of "union" on IRanges objects (yes, you could say that `union()` and all the set-ops are broken on IRanges objects). You would need to use something like `c(levels(rel1), levels(rel2)[!(levels(rel2) %in% levels(rel1))])` instead. Just some thought. 

H.

**Comment by LTLA** [comment]
Thanks Herve. I hadn't done a lot of testing with non-`Vector` "levels", that's probably why they're not happy. I'll have a closer look later.

I too had considered the relationship to factors. In fact, I had considered implementing a `Factor` class to put into _S4Vectors_, which would effectively negate the need for this package.  The entire `IndexedRelations` class would basically be a `DataFrame` of `Factor`s, which would natively support comparisons thanks to Bioconductor/S4Vectors#36.

I didn't do it because it was too much effort to get buy-in from everyone, but I'm happy to reconsider whether a `Factor` class would be a good idea. We would then have stuff like `GRangesFactor` and `IRangesFactor` and `DataFrameFactor`, for example. This would involve cannabilizing some of the code in this package, or possibly just renaming this package to `S4Factors` and stripping out the multi-dimensional component.

**Comment by hpages** [comment]
A Factor class sounds good. Even if it goes to S4Vectors (but I don't think it needs to for now), I don't think it would negate the need for IndexedRelations. Maybe you'd still want to have the IndexedRelations class instead of using a naked DataFrame with Factor columns. The former lets you define a specific validity method (all the columns in the DataFrame must be Factor objects). But also, and most importantly, it allows you to stay away from the schizophrenic nature of DataFrame objects. The main question is: do you want `length(x)`, `x[i]`, and `mcols(x)` to behave like now, or do you want them to return the number of columns (for `length(x)`), to subset along the columns (for `x[i]`), and to access metadata columns that run along the columns (for `mcols(x)`), which is what using a naked DataFrame would give us. In other words, should the backbone of an IndexedRelations creature run vertically or horizontally? I don't know the details of your use cases but I suspect you want the former. In this case, IndexedRelations should not extend DataFrame either but have a slot containing a DataFrame (or just a list of Factor objects of the same length, because it would be easy and cheap to have the `partners()` getter wrap this into a DataFrame on-the-fly). That is, composition should be used rather than inheritance.

This would also mean that duplicated levels wouldn't be allowed right? (They are deprecated for ordinary factors.)

Note that if you go the Factor route, we should be able to extract individual Factors from an IndexedRelations object `x`. One approach is to support 2D-style subsetting i.e. `x[i, j]`. Then `x[ , j]` would extract the j-th Factor. This means that IndexedRelations objects are now 2D objects so they support `dim()`, `nrow()`, `ncol`, `dimnames()`, `rownames()`, and `colnames()`. But unlike data-frame-like objects, their backbone is still running along their 1st dimension so `length(x)` and `names(x)` are the same as `nrow(x)` and `rownames(x)`, respectively. This is what SummarizedExperiment objects do!

Note that `x[i, j]` would only return an individual factor when `j` is a single integer and `drop=TRUE` (which is the default). `x[i, j, drop=FALSE]` should always behave like an _endomorphism_ i.e. return another IndexedRelations object.

Throwing around another possible name for this class/package: what about "Interactions"? If you look at the man page for `base::split()`, `f` can be a factor or a list of factors "in which case their **interaction** is used for the grouping". I think there are other places in base R where the term "interaction" is used in a similar context. And the "interaction" between factors of the same length can actually be computed with `interaction()` (which returns another factor of the same length). I'm not saying that we should try to implement an `interaction()` method for Factor objects (it probably wouldn't make a lot of sense), just that the term "interaction" has some roots in base R that match your IndexedRelations concept.

Also what about `index()` and `levels()` instead of `partners()` and `featureSets()`? `index(x)[[j]]` and `levels(x)[[j]]` would be the same as `as.integer(x[ , j])` and `levels(x[ , j])`, respectively.

Note that it's not clear what coercion to DataFrame should do: produce columns of Factor objects or produce "decoded columns" i.e. columns obtained with `levels(x)[[j]][index(x)[[j]]]` like now. Sounds like the latter might be more useful, maybe. One thing that could be useful to have is a standard way to "decode" (or "unfactor") a factor object. In base R, one can use `as.vector()` or `as.character()` for this. But for Factor objects we need something else (and we can't use `as( , "Vector")` either). Maybe we could introduce `decode()` (or `unfactor()` or `defactor()`) for this? So for example the user could easily extract the j-th column in decoded form with `decode(x[ , j])`. `decode()` could also work directly on IndexedRelations object `x` and return the DataFrame of decoded columns. In that case, we should probably have `as(x, "DataFrame")` return a DataFrame with Factor columns, rather than being equivalent to `decode(x)`.

**Comment by LTLA** [comment]
> In other words, should the backbone of an IndexedRelations creature run vertically or horizontally? I don't know the details of your use cases but I suspect you want the former.

Yes.

> That is, composition should be used rather than inheritance.

Yes.

> This would also mean that duplicated levels wouldn't be allowed right?

Right. We might as well be consistent with expected behaviour for base factors.

> Throwing around another possible name for this class/package: what about "Interactions"?

I don't have any objections to that.

> Also what about `index()` and `levels()` instead of `partners()` and `featureSets()`?

Yes. That was one of the major appeals of a hypothetical `Factor` class. I could just re-use `levels()`, for example, and that would make my life a lot easier.

> Sounds like the latter might be more useful, maybe.

I would agree, as people probably just want some easily digestible columns if they're coercing. But I'm open to being convinced the other way, as power users would probably want to keep factors.

**Comment by hpages** [comment]
Excellent. If you rename the package, you'll need to resubmit. Not a big deal. We'll re-assign the package to me and we'll close this issue. Looking forward to what you'll come up with. Thanks!

**Comment by LTLA** [comment]
Having thought about this further, I'm starting to (re-)think that implementing `Factor` would render `IndexedRelations` itself unnecessary. This is strictly from my own developer perspective; I developed `IndexedRelations` as a base for `GenomicInteractions`, but if I have `GRangesFactor`s available, I could just set them as individual `parallelSlotNames` in the `GenomicInteractions` definition and not require derivation of `GenomicInteractions` from `IndexedRelations`. Even a more esoteric use case like kevinrue/unisets#47 could just use individual `DataFrameFactor`s as parallel slots. 

An `IndexedRelations` class would only be necessary if the number of partners was variable. This seems like a pretty unusual scenario to me. Well, perhaps someone might need it, but I don't.

**tl;dr** I'm thinking of just stripping it all the way down to an _S4Factors_ package. Throw in some `DataFrameFactor` and `GRangesFactor` definitions, maybe. Possibly migrate a specialized `findOverlaps` method that uses the levels to perform the overlaps quickly.

**Comment by hpages** [comment]
I see. I thought the motivation behind IndexedRelations was to support an arbitrary number of Factors (and I was actually wondering/curious about what use case you had in mind for this). But if the only use case is a pair of Factor, maybe IndexedRelations is indeed not longer justified.
Note that it would be weird to have the Factor class and its specializations like GRangesFactor, DataFrameFactor, etc... all in one package. It sounds more sensible to me to have Factor in S4Vectors, and the specialization classes in the same package where the levels class is defined. So DataFrameFactor also in S4Vectors, but GRangesFactor, in GenomicRanges, etc...

**Comment by LTLA** [comment]
> Note that it would be weird to have the Factor class and its specializations like GRangesFactor, DataFrameFactor, etc... all in one package.

I'm glad that you think this way, because I didn't want to do it.

**Comment by hpages** [comment]
No problem, I'll add this to S4Vectors and GenomicRanges in the next few days:
- https://github.com/Bioconductor/S4Vectors/issues/38
- https://github.com/Bioconductor/GenomicRanges/issues/25

**Comment by LTLA** [comment]
I'm going to close this and continue discussions in the two links above.
